cmd/installer/cli/headless/install/orchestrator_test.go

```diff
@@ -0,0 +1,233 @@
+				return nil, &apitypes.APIError{
+					StatusCode: 400,
+					Message:    "field errors",
+					Errors: []*apitypes.APIError{
+						{
+							Field:   "database_host",
+							Message: "required field missing",
+						},
+						{
+							Field:   "replica_count",
+							Message: "value \"10\" exceeds maximum allowed value 5",
+						},
+						{
+							Field:   "enable_ssl",
+							Message: "validation rule failed: SSL requires cert_path to be set",
+						},
+					},
+				}
+			},
+			configValues: apitypes.AppConfigValues{
+				"database_host": apitypes.AppConfigValue{
+					Value: "",
+				},
+				"replica_count": apitypes.AppConfigValue{
+					Value: "10",
+				},
+			},
+			expectError:         true,
+			expectedErrorMsg:    "application configuration validation failed: field errors:\n  - Field 'database_host': required field missing\n  - Field 'replica_count': value \"10\" exceeds maximum allowed value 5\n  - Field 'enable_ssl': validation rule failed: SSL requires cert_path to be set",
+			expectedLogMessages: []string{},
+			expectedProgressMessages: []string{
+				"Configuring application...",
+				"Application configuration failed",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Setup
+			mockClient := &mockAPIClient{
+				patchLinuxInstallAppConfigValuesFunc: tt.mockPatchFunc,
+			}
+
+			// Create logger with capture hook
+			logger := logrus.New()
+			logger.SetOutput(io.Discard) // Discard actual output, we only want the hook
+			logCapture := newLogMessageCapture()
+			logger.AddHook(logCapture)
+
+			// Capture progress writer messages
+			progressCapture := newProgressMessageCapture()
+
+			orchestrator := &orchestrator{
+				apiClient:      mockClient,
+				target:         apitypes.InstallTargetLinux,
+				progressWriter: progressCapture.Writer(),
+				logger:         logger,
+			}
+
+			opts := HeadlessInstallOptions{
+				ConfigValues: tt.configValues,
+			}
+
+			// Execute
+			err := orchestrator.configureApplication(context.Background(), opts)
+
+			// Assert error
+			if tt.expectError {
+				require.Error(t, err)
+				if tt.expectedErrorMsg != "" {
+					assert.Equal(t, tt.expectedErrorMsg, err.Error())
+				}
+			} else {
+				require.NoError(t, err)
+			}
+
+			// Assert log messages
+			assert.Equal(t, tt.expectedLogMessages, logCapture.Messages(), "log messages should match")
+
+			// Assert progress messages
+			assert.Equal(t, tt.expectedProgressMessages, progressCapture.Messages(), "progress messages should match")
+		})
+	}
+}
+
+// logMessageCapture is a logrus hook that captures log messages for testing.
+// It implements the logrus.Hook interface to intercept all log messages and
+// store them for verification in tests. This allows tests to verify that
+// specific log messages are produced without requiring actual log output.
+type logMessageCapture struct {
```

Comment:
this seems helpful overall, should it be moved to a util package so it can be re-used? maybe in the spinner package even?

Reply:
i will reuse it heavily within this package. if we want to move it to a utility package perhaps we can do that later.